src/util/TunnelIPv6Interface.cpp

```diff
@@ -148,7 +182,24 @@ TunnelIPv6Interface::on_address_added(const struct in6_addr &address, uint8_t pr
 	syslog(LOG_INFO, "TunnelIPv6Interface: \"%s/%d\" was added to \"%s\"", in6_addr_to_string(address).c_str(), prefix_len,
 	       get_interface_name().c_str());
 
-	mAddressWasAdded(address, prefix_len);
+	mUnicastAddressWasAdded(address, prefix_len);
+}
+
+void
+TunnelIPv6Interface::on_multicast_address_joined(const struct in6_addr &address)
+{
+	if (mPendingMulticastAddresses.count(address)) {
+		mPendingMulticastAddresses.erase(address);
+	}
+
+	syslog(LOG_INFO, "TunnelIPv6Interface: \"%s\" was added to \"%s\"", in6_addr_to_string(address).c_str(),
+	       get_interface_name().c_str());
+
+	if (mMulticastAddresses.count(address) == 0) {
+		// update new addresses only
+		mMulticastAddresses.insert(address);
+		mMulticastAddressWasJoined(address);
```

Comment:
I think for unicast address we emit the signal on any address add (we don't maintain a list of unicast address in this class and check if it's already in list or not from this class before emitting `mUnicastAddressWasAdded`). 

Can we consider doing the same for multicast addresses? 
If we go for this, I think we won't need the `mMulticastAddresses`  (we just need a pending list of multicast addresses to add which would be simpler).

Note that the `NCPInstanceBase` class itself will manage all address types (and handle the `mMulticastAddressWasJoined` signal. It maintains list of multicast addresses and it would handle if address is already in the list. 

Thoughts?

Reply:
The problem I have intended to solve here is that because we are joining all the multicast groups the NCP has joined by itself (at line 663 in the same file). This will also trigger a MLD packet which will also be captured by the listener. The `mMulticastAddresses` has been added to prevent joining the group again.

Reply:
As you say, OpenThread itself already joins the same multicast address and `wpantund` would sync addresses between OT/NCP and host interface. I think the addition of the same address in `wpantund` (line 663) is actually redundant and can be possibly removed. This would then simplify the code (remove the need for `mMulticastAddresses` in this class).

Thoughts?

Reply:
The addition of the same address at line 663 is required to notify the linux kernel to not filter these addresses.

Reply:
Let me first clarify my earlier comment. I realized we may have been talking about two different things :)...I initially thought your were referring to this line (where from constructor we join a multicast group eventhough OT/NCP itself would join the same address):

https://github.com/openthread/wpantund/blob/master/src/wpantund/NCPInstanceBase.cpp#L149-L151

This is what I refereed to as redundant in my earlier comment - not line 663.
I think we can maybe address this later/different PR.

---

Regarding multicast addresses originating from NCP and added on linux interface, I still think we don't need a separate `mMulticastAddresses` in `TunnelIPv6Interface` class.

The `NCPInstanceBase` class already keeps track of all multicast addresses and also  tracks their origin (who added the address, NCP, interface, user). The `TunnelIPv6Interface`  does not need to track/save list of addresses, it just needs to provide API to add/remove addresses and signals to notify when an address is added/removed (from interface). The `NCPInstanceBase` would not re-add an address again if it's already in its list. 
- NCP reports a new multicast address
- `NCPInstanceBase` would add it in its list (marking origin as NCP) and then request it to be added on linux interface (calls `TunnelIPv6Interface::join_multicast_address()`.
- A bit later (when address is actually added on interface), we get it in `TunnelIPv6Interface`  through MLD. 
- At this point, it is ok for `TunnelIPv6Interface` to invoke `mMulticastAddressWasJoined` signal (which is hooked to `NCPInstanceBase::multicast_address_was_joined()`).
- The `mlticast_address_was_joined()` would first check if the address is new or not (if it's already in its list) and takes no action.
https://github.com/openthread/wpantund/blob/master/src/wpantund/NCPInstanceBase-Addresses.cpp#L767-L792

This is same pattern we use for unicast addresses already. My suggestion was to do the same for multi-cast (unless there is clear difference between unicast/multicast...) If we can go for this we can `mMulticastAddresses` in `TunnelIPv6Interface` and make it simpler.

Thoughts?

Reply:
I've tried to remove mMulticastAddresses but it caused the OpenThread toranj test to fail. I'll investigate what's going wrong.

Reply:
Can you please add which tests are failing (with removal of `mMulticastAddresses`)?

Reply:
It's test-02 test-join which reports that the device is not in offline state when `wpanctl leave` is called. Waiting for about a second suffices to put the device to the right state.

Reply:
On my workstation it's test-002-form but it works fine on the travis ci. I'll double check what's going wrong.